sns_ik_lib/include/sns_ik/sns_vel_ik_base.hpp

```diff
@@ -0,0 +1,191 @@
+   * This method will update the number of joints in the solver
+   * Requirements: inputs must be the same size and dqLow < dqUpp
+   * @param dqLow: lower bound on the velocity of each joint
+   * @param dqUpp: upper bound on the velocity of each joint
+   * @return: true iff successful
+   */
+  bool setVelBnd(const Eigen::ArrayXd& dqLow, const Eigen::ArrayXd& dqUpp);
+
+  /**
+   * Solve a velocity IK problem with no null-space bias of joint-space optimization.
+   *
+   *  This method implements   "Algorithm 1: SNS algorithm"   from the paper:
+   *  "Control of Redundant Robots Under Hard Joint Constraint: Saturation in the Null Space"
+   *   by: Fabrizio Flacco, Alessandro De Luca, Oussama Khatib
+   *
+   * Solve for joint velocity dq and task scale s:
+   *
+   *  maximize: s
+   *  subject to:
+   *    s * dx = J * dq
+   *    0 < s <= 1
+   *    dqLow <= dq <= dqUpp      ( bounds set in constructor or setVelBnd() )
+   *
+   * @param J: Jacobian matrix, mapping from joint to task space. Size = [nTask, nJoint]
+   * @param dx: task velocity vector. Length = nTask
+   * @param[out] dq: joint velocity solution. Length = nJoint
+   * @param[out] taskScale: task scale.  fwdKin(dq) = taskScale*dx
+   *                            taskScale == 1.0  --> task was feasible
+   *                            taskScale < 1.0  --> task was infeasible and had to be scaled
+   * @return: Success: the algorithm worked correctly and satisfied the problem statement
+   *          otherwise: something went wrong
+   */
+  ExitCode solve(const Eigen::MatrixXd& J, const Eigen::VectorXd& dx,
+                      Eigen::VectorXd* dq, double* taskScale);
+
+
+protected:
+
+  /*
+   * protected constructor: require factory method to create an object.
+   */
+  SnsVelIkBase(int nJnt) : nJnt_(nJnt), dqLow_(nJnt), dqUpp_(nJnt) {};
+
+  /*
+   * Check that dqLow_ <= dq <= dqUpp_
+   * @param dq: joint velocity to test
+   * @return: true iff qLow <= q <= qUpp
+   *          if dq.empty() return false
+   */
+  bool checkVelBnd(const Eigen::VectorXd& dq);
+
+  /*
+   * Solve the following equation for the variable dq:
+   *    J * W * (dq - dqNull) = dx - J*dqNull
+   * This equation appears throughout the SNS-IK papers. One example is in block "D" of Algorithm 1:
+   *  "Control of Redundant Robots Under Hard Joint Constraint: Saturation in the Null Space"
+   *   by: Fabrizio Flacco, Alessandro De Luca, Oussama Khatib
+   *
+   * PRECONDITION:
+   * --> Assumes that linSolver_ has been initialized with J*W
+   *
+   * @param J: Jacobian matrix, mapping from joint to task space. Size = [nTask, nJoint]
+   * @param JW: J*W = Jacobian projected onto the active joints. Size = [nTask, nJoint]
+   * @param dqNull: null-space joint velocity. Size = nJoint
+   * @param dx: task velocity vector. Length = nTask
+   * @param[out] dq: joint velocity solution. Length = nJoint
+   * @param[out, opt] resErr: residual error (norm-squared)
+   * @return: true --> success!
```

Comment:
I'm a bit confused between the usage of `ExitCode` enums and `bool`s as return values. Would it make sense to be consistent in return values?

Reply:
The `ExitCode` is primarily used for external return values, or places where the calling code might need to know the cause of failure. I suppose that `solveProjectionEquation` should probably also use ExitCode, in case an external user wanted to call it. The main reason to use a bool return where possible is that it is easier for an external user who does not want to create an instance of a special enum type to check the result - a bool check is easier.

Reply:
I agree a bool check is easier, but consistency is more important if this is intended to be an external facing function. I do agree the enum is cumbersome (as I mentioned above).